File: services/graphql-dms/src/mutation/customer_tracking_links.rs
use async_graphql::{Context, InputObject, Object};
use uuid::Uuid;

#[derive(Debug, Clone, InputObject)]
pub struct CreateCustomerTrackingLinkInput {
    pub delivery_task_id: Uuid,
    pub tracking_token: String,
    pub is_active: Option<bool>,
    pub expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Default)]
pub struct Mutation;

#[Object(name = "DmsCustomerTrackingLinksMutations")]
impl Mutation {
    async fn create_customer_tracking_link(
        &self,
        ctx: &Context<'_>,
        payload: CreateCustomerTrackingLinkInput,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
            "insert into dms.customer_tracking_links (delivery_task_id, tracking_token, is_active, expires_at) values ($1, $2, $3, $4) returning *"
        )
        .bind(payload.delivery_task_id)
        .bind(payload.tracking_token)
        .bind(payload.is_active)
        .bind(payload.expires_at)
        .fetch_one(db)
        .await?)
    }

    async fn update_customer_tracking_link_delivery_task_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        delivery_task_id: Uuid,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
            "update dms.customer_tracking_links set delivery_task_id = $1 where id = $2 returning *",
        )
        .bind(delivery_task_id)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_customer_tracking_link_tracking_token(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        tracking_token: String,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
            "update dms.customer_tracking_links set tracking_token = $1 where id = $2 returning *",
        )
        .bind(tracking_token)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_customer_tracking_link_is_active(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        is_active: Option<bool>,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(
            sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
                "update dms.customer_tracking_links set is_active = $1 where id = $2 returning *",
            )
            .bind(is_active)
            .bind(id)
            .fetch_one(db)
            .await?,
        )
    }

    async fn update_customer_tracking_link_access_count(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        access_count: Option<i32>,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
            "update dms.customer_tracking_links set access_count = $1 where id = $2 returning *",
        )
        .bind(access_count)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_customer_tracking_link_last_accessed_at(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        last_accessed_at: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
            "update dms.customer_tracking_links set last_accessed_at = $1 where id = $2 returning *",
        )
        .bind(last_accessed_at)
        .bind(id)
        .fetch_one(db)
        .await?)
    }

    async fn update_customer_tracking_link_expires_at(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        expires_at: Option<chrono::DateTime<chrono::Utc>>,
    ) -> async_graphql::Result<crate::models::customer_tracking_links::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;
        Ok(
            sqlx::query_as::<_, crate::models::customer_tracking_links::Model>(
                "update dms.customer_tracking_links set expires_at = $1 where id = $2 returning *",
            )
            .bind(expires_at)
            .bind(id)
            .fetch_one(db)
            .await?,
        )
    }

    async fn remove_customer_tracking_link(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
    ) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;
        let result = sqlx::query("delete from dms.customer_tracking_links where id = $1")
            .bind(id)
            .execute(db)
            .await?;
        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new(
                "Unable to delete customer tracking link",
            ));
        }
        Ok("Customer tracking link removed successfully".into())
    }
}

File: src/graphql/mutations/dms/customerTrackingLinks.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// CUSTOMER TRACKING LINK MUTATIONS
// ============================================================================

/**
 * Creates a new customer tracking link.
 * @param payload The input data for creating the customer tracking link.
 */
export const createCustomerTrackingLink = graphql(`
  mutation CreateCustomerTrackingLink($payload: CreateCustomerTrackingLinkInput!) {
    dms {
      createCustomerTrackingLink(payload: $payload) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the delivery task ID of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param deliveryTaskId The new delivery task ID.
 */
export const updateCustomerTrackingLinkDeliveryTaskId = graphql(`
  mutation UpdateCustomerTrackingLinkDeliveryTaskId($id: UUID!, $deliveryTaskId: UUID!) {
    dms {
      updateCustomerTrackingLinkDeliveryTaskId(id: $id, deliveryTaskId: $deliveryTaskId) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the tracking token of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param trackingToken The new tracking token.
 */
export const updateCustomerTrackingLinkTrackingToken = graphql(`
  mutation UpdateCustomerTrackingLinkTrackingToken($id: UUID!, $trackingToken: String!) {
    dms {
      updateCustomerTrackingLinkTrackingToken(id: $id, trackingToken: $trackingToken) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the active status of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param isActive The new active status.
 */
export const updateCustomerTrackingLinkIsActive = graphql(`
  mutation UpdateCustomerTrackingLinkIsActive($id: UUID!, $isActive: Boolean) {
    dms {
      updateCustomerTrackingLinkIsActive(id: $id, isActive: $isActive) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the access count of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param accessCount The new access count.
 */
export const updateCustomerTrackingLinkAccessCount = graphql(`
  mutation UpdateCustomerTrackingLinkAccessCount($id: UUID!, $accessCount: Int) {
    dms {
      updateCustomerTrackingLinkAccessCount(id: $id, accessCount: $accessCount) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the last accessed at timestamp of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param lastAccessedAt The new last accessed at timestamp.
 */
export const updateCustomerTrackingLinkLastAccessedAt = graphql(`
  mutation UpdateCustomerTrackingLinkLastAccessedAt($id: UUID!, $lastAccessedAt: DateTime) {
    dms {
      updateCustomerTrackingLinkLastAccessedAt(id: $id, lastAccessedAt: $lastAccessedAt) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Updates the expiration timestamp of a customer tracking link.
 * @param id The UUID of the customer tracking link to update.
 * @param expiresAt The new expiration timestamp.
 */
export const updateCustomerTrackingLinkExpiresAt = graphql(`
  mutation UpdateCustomerTrackingLinkExpiresAt($id: UUID!, $expiresAt: DateTime) {
    dms {
      updateCustomerTrackingLinkExpiresAt(id: $id, expiresAt: $expiresAt) {
        id
        trackingToken
        isActive
        accessCount
        lastAccessedAt
        expiresAt
        createdAt
        updatedAt
        deliveryTask {
          id
          deliveryAddress
          recipientName
        }
      }
    }
  }
`);

/**
 * Removes a customer tracking link by its ID.
 * @param id The UUID of the customer tracking link to remove.
 */
export const removeCustomerTrackingLink = graphql(`
  mutation RemoveCustomerTrackingLink($id: UUID!) {
    dms {
      removeCustomerTrackingLink(id: $id)
    }
  }
`);
